README.md
# CNN Google News

Node.js app for generating the feed required by
[Google News](https://play.google.com/store/newsstand) for displaying
[CNN](https://cnn.com) content in the Google News app on iOS and Android.

[Google News Docs](https://support.google.com/news/producer/answer/6170026?hl=en)

## Getting Started

```shell
$ npm install
$ npm start
```

Go to http://localhost:5000

# How to deploy to harbor
This project builds through Codeship. Any commit on develop/master will deploy to Harbor through Codeship + Bamboo.


## Licensing
See [LICENSE.md](./LICENSE.md) for details.


CHANGELOG.md
# CNN Google Newsstand Changelog

## 2018-09-27, Version 1.14.0, @khowen

### Notable changes
- adding a business feed
- updated npm token usage

### Commits
* [[`aaa47b89cb`](https://github.com/jamsyoung/cnn-google-newsstand/commit/aaa47b89cb)] - **(SEMVER-MINOR)** \[Ds-536\]: add new feed (Katie Holland Owen) [#72](https://github.com/jamsyoung/cnn-google-newsstand/pull/72)

## 2018-06-26, Version 1.13.0, @khowen

### Notable changes
- update README

### Commits
* [[`bb5936cfca`](https://github.com/jamsyoung/cnn-google-newsstand/commit/bb5936cfca)] - **(SEMVER-MINOR)** **README**: update README (Katie Holland Owen) [#66](https://github.com/jamsyoung/cnn-google-newsstand/pull/66)

## 2018-02-12, Version 1.12.3, @khowen

### Notable changes

- adding fix for missing video size

### Commits
* [[`2d4a085254`](https://github.com/jamsyoung/cnn-google-newsstand/commit/2d4a085254)] - fixing conflict (Sean Joseph)
* [[`6dec9bd9ec`](https://github.com/jamsyoung/cnn-google-newsstand/commit/6dec9bd9ec)] - **(SEMVER-MINOR)** **feed-generator**: add video sizing (Katie Owen) [#60](https://github.com/jamsyoung/cnn-google-newsstand/pull/60)




## 2018-02-09, Version 1.12.2, @sjoseph

### Notable changes

- adding npm key

### Commits
>>>>>>> master
* [[`b5f58a10c4`](https://github.com/jamsyoung/cnn-google-newsstand/commit/b5f58a10c4)] - adding npm key (Sean Joseph)




## 2018-02-09, Version 1.12.1, @sjoseph

### Notable changes
- Added Winter Olympics 2018 feed
- Updated content retriever version 3.0.0

### Commits
* [[`c15be80b37`](https://github.com/jamsyoung/cnn-google-newsstand/commit/c15be80b37)] - CNNGNS-100 & CNNGNS-101 Winter Olympics and Content Retriever (#58) (sean joseph) [#58](https://github.com/cnnlabs/cnn-google-newsstand/pull/58)




## 2017-10-24, Version 1.12.0, @khowen

### Notable changes
- Add blacklisted section

### Commits
* [[`1b4a461490`](https://github.com/jamsyoung/cnn-google-newsstand/commit/1b4a461490)] - **(SEMVER-MINOR)** **config**: add blacklisted section (Katie Owen) [#51](https://github.com/jamsyoung/cnn-google-newsstand/pull/51)




## 2017-8-14, Version 1.11.1, @khowen

### Notable changes
- Update style section regex for style URL changes

### Commits
* [[`9b9da8807c`](https://github.com/jamsyoung/cnn-google-newsstand/commit/9b9da8807c)] - **(SEMVER-MINOR)** **google-newsstand-articles**: update style URL regex (Katie Owen) [#43](https://github.com/jamsyoung/cnn-google-newsstand/pull/43)




## 2017-6-22, Version 1.10.1, @zoitec

### Notable changes
- fix to allow videos


* [[`af8de9c9a1`](https://github.com/jamsyoung/cnn-google-newsstand/commit/af8de9c9a1)] - fix to allow videos (Sean Joseph)




## 2017-6-09, Version 1.10.0, @zoitec

### Notable changes
- Adding video template type for analytics team




* [[`c4728da4e6`](https://github.com/jamsyoung/cnn-google-newsstand/commit/c4728da4e6)] - CNNGNS-85 Setting correct templateType based on the presence of a video (#35) (sean joseph)




## 2017-6-09, Version 1.9.1, @zoitec

### Notable changes
- Fixing gallery caption bug

* [[`cb5dee660d`](https://github.com/jamsyoung/cnn-google-newsstand/commit/cb5dee660d)] - CNNGNS-48 Adding enable flag and fix production error when gallery caption is undefined (#34) (sean joseph)




## 2017-6-08, Version 1.9.0, @zoitec

### Notable changes
- Fixing gallery footer and media image captions

* [[`cfa6a3262d`](https://github.com/jamsyoung/cnn-google-newsstand/commit/cfa6a3262d)] - CNNGNS-48 Inline Galleries in GNS (#33) (sean joseph)




## 2017-5-22, Version 1.8.0, @zoitec

### Notable changes
- Fixing gallery footer and media image captions


* [[`7444983f78`](https://github.com/jamsyoung/cnn-google-newsstand/commit/7444983f78)] - fixing merge conflicts (Sean Joseph)
* [[`ee8c673b4a`](https://github.com/jamsyoung/cnn-google-newsstand/commit/ee8c673b4a)] - CNNGNS-87 fix double captions (#30) (sean joseph)
* [[`8ffee7ba82`](https://github.com/jamsyoung/cnn-google-newsstand/commit/8ffee7ba82)] - fixing gallery image (Sean Joseph)
* [[`0a7c126932`](https://github.com/jamsyoung/cnn-google-newsstand/commit/0a7c126932)] - CNNGNS-61 - using correct footer go galleries (#29) (sean joseph)


## 2017-5-19, Version 1.7.9, @zoitec

### Notable changes
- Adding new Style and Travel section


* [[`7b954392b0`](https://github.com/jamsyoung/cnn-google-newsstand/commit/7b954392b0)] - CNNGNS-87 CNNGNS-88 Add Style and Travel Section (#28) (sean joseph)




## 2017-5-18, Version 1.7.8, @zoitec

### Notable changes
- Fix Stale Tech section


* [[`b96d6435a9`](https://github.com/jamsyoung/cnn-google-newsstand/commit/b96d6435a9)] - Feature/cnngns 61 (#27) (sean joseph)




## 2017-5-18, Version 1.7.7, @zoitec

### Notable changes
- Fix markup for Gallery RSS Items


* [[`6d63737a94`](https://github.com/jamsyoung/cnn-google-newsstand/commit/6d63737a94)] - CNNGNS- fix gallery rss items (#26) (sean joseph)




## 2017-5-17, Version 1.7.6, @zoitec

### Notable changes
- subscribe to cnn.gallery queue messages

* [[`ff26620a47`](https://github.com/jamsyoung/cnn-google-newsstand/commit/ff26620a47)] - CNNGNS-61 Add gallery to the allow list (#25) (sean joseph)




## 2017-5-15, Version 1.7.5, @zoitec

### Notable changes
- subscribe to cnn.gallery queue messages

* [[`60e2b6e7fa`](https://github.com/jamsyoung/cnn-google-newsstand/commit/60e2b6e7fa)] - CNNGNS-61 - Stand Alone Galleries in GNS (#18) (sean joseph)



## 2017-5-15, Version 1.7.4, @zoitec

### Notable changes
- Enable stand alone galleries

* [[`60e2b6e7fa`](https://github.com/jamsyoung/cnn-google-newsstand/commit/60e2b6e7fa)] - CNNGNS-61 - Stand Alone Galleries in GNS (#18) (sean joseph)




## 2017-5-4, Version 1.7.3, @zoitec

### Notable changes
- Filter out live blogs


### Commits
* [[`1328bd180c`](https://github.com/jamsyoung/cnn-google-newsstand/commit/1328bd180c)] - filtering out articles that are non type of story.. ie live blog (#23) (sean joseph)




## 2017-4-3, Version 1.7.2, @zoitec

### Notable changes

config.js
/*
 * Copyright 2016 Turner Broadcasting System, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';


const nconf = require('nconf');



// whitelist environment variables
nconf.env([
    'BUSINESS_SECTIONS',
    'CLOUDAMQP_AUTH',
    'DYNAIMAGE_AUTH',
    'ENVIRONMENT',
    'GNS_BLACK_LIST',
    'GNS_GENERIC_THUMB_IMAGE',
    'GNS_TASK_INTERVAL_MS',
    'GNS_ENABLE_ELECTION_STORY',
    'GNS_ELECTION_MODULE_TEST',
    'GNS_ELECTION_TITLE',
    'GNS_ELECTION_IMG_ENV',
    'GNS_ELECTION_SLUG',
    'GNS_ELECTION_CAPTION',
    'GNS_ELECTION_STORY_CONSTANT_UPDATE',
    'GNS_ELECTION_STORY_CONSTANT_UPDATE_URL',
    'GNS_ELECTION_LOWER_HEADER',
    'GNS_ELECTION_MODULE_LINK_1',
    'GNS_ELECTION_MODULE_LINK_2',
    'GNS_ENABLE_INLINE_GALLERIES',
    'GNS_MONITORING_TEST',
    'HYPATIA_TIMEOUT',
    'HYPATIA_HOST',
    'ACCESS_KEY_ID',
    'SECRET_ACCESS_KEY',
    'PORT',
    'LOGZIO_TOKEN'
]);



// These are required to be set to start up
if (!nconf.get('ENVIRONMENT') || !nconf.get('PORT') || !nconf.get('CLOUDAMQP_AUTH') || !nconf.get('LOGZIO_TOKEN') || !nconf.get('ACCESS_KEY_ID') || !nconf.get('SECRET_ACCESS_KEY') || !nconf.get('DYNAIMAGE_AUTH')) {
    console.error('ENVIRONMENT, PORT, CLOUDAMQP_AUTH, LOGZIO_TOKEN, ACCESS_KEY_ID, DYNAIMAGE_AUTH and/or SECRET_ACCESS_KEY are not set');
    process.exit(1);
}

let blackList = '\\/studentnews\\/,\\/videos\\/spanish\\/,fast-facts\\/index.html$,\\/applenews-live.*,\\/cnn.com\\/\\d{4}\\/\\d{2}\\/\\d{2}\\/cnn-info,\\/vr\\/,\\/cnn-underscored\\/',
    config = {
        default: {
            cloudamqpConnectionString: `amqp://${nconf.get('CLOUDAMQP_AUTH')}@red-rhino.rmq.cloudamqp.com/cnn-towncrier`,
            gnsBlackList: (nconf.get('GNS_BLACK_LIST')) ? nconf.get('GNS_BLACK_LIST') : blackList,
            gnsTaskIntervalMS: (nconf.get('GNS_TASK_INTERVAL_MS')) ? parseInt(nconf.get('GNS_TASK_INTERVAL_MS')) : 1000 * 60 * 30, // 30 minutes
            lsdHosts: 'lsd-prod-pub-cop.turner.com,lsd-prod-pub-56m.turner.com',
            exchangeName: 'cnn-town-crier-ref',
            queueNameArticles: `cnn-google-newsstand-articles-${nconf.get('ENVIRONMENT').toLowerCase()}`,
            queueNameVideos: `cnn-google-newsstand-videos-${nconf.get('ENVIRONMENT').toLowerCase()}`,
            // TODO: remove when Money feed is decommissioned
            routingKeysArticles: ['cnn.article', 'money.article', 'cnn.gallery'],
            routingKeysVideos: ['cnn.video'],
            adbpTrackingURL: 'https://smetrics.cnn.com/b/ss/cnnoffsitedev',
            gnsTurnOnElectionModule: (nconf.get('GNS_ENABLE_ELECTION_STORY')) ? nconf.get('GNS_ENABLE_ELECTION_STORY') : false,
            gnsElectionModuleTest: (nconf.get('GNS_ELECTION_MODULE_TEST')) ? nconf.get('GNS_ELECTION_MODULE_TEST') : false,
            gnsElectionSlug: (nconf.get('GNS_ELECTION_SLUG')) ? nconf.get('GNS_ELECTION_SLUG') : 'what-is-the-hatch-act',
            gnsElectionCaption: (nconf.get('GNS_ELECTION_CAPTION')) ? nconf.get('GNS_ELECTION_CAPTION') : 'A candidate needs 270 Electoral College votes to win the presidency. These modules show the total Electoral College votes and when a state has a projected winner (Maine and Nebraska allow Electoral College votes to be split). Not all candidates are listed. All party representation can be seen in the full results. CNN will broadcast a projected winner only after an extensive review of data from a number of sources.',
            gnsElectionTitle: (nconf.get('GNS_ELECTION_TITLE')) ? nconf.get('GNS_ELECTION_TITLE') : '<h3 class="style-id:presResultsHeader"><span class="style-id:presResultsHeader">Presidential Results</span></h3>',
            gnsElectionStoryConstantUpdate: (nconf.get('GNS_ELECTION_STORY_CONSTANT_UPDATE')) ? nconf.get('GNS_ELECTION_STORY_CONSTANT_UPDATE') : false,
            gnsElectionStoryConstantUpdateURL: (nconf.get('GNS_ELECTION_STORY_CONSTANT_UPDATE_URL')) ? nconf.get('GNS_ELECTION_STORY_CONSTANT_UPDATE_URL') : 'http://www.cnn.com/2016/10/28/politics/trump-bus-denmark-trnd/index.html',
            gnsElectionModuleLowerHeader: (nconf.get('GNS_ELECTION_LOWER_HEADER')) ? nconf.get('GNS_ELECTION_LOWER_HEADER') : 'Election Day Highlights',
            gnsElectionModuleLink1: (nconf.get('GNS_ELECTION_MODULE_LINK_1')) ? nconf.get('GNS_ELECTION_MODULE_LINK_1') : '<p class="style-id:electionLinks"><a class="style-id:electionLink" href="http://www.cnn.com/election/results">Full&nbsp;Election&nbsp;Results</a> | <a class="style-id:electionLink" href="http://www.cnn.com/election/president">Presidential</a> | <a class="style-id:electionLink" href="http://www.cnn.com/election/senate">Senate</a></p>',
            gnsElectionModuleLink2: (nconf.get('GNS_ELECTION_MODULE_LINK_2')) ? nconf.get('GNS_ELECTION_MODULE_LINK_2') : '<p class="style-id:electionLinks2"><a class="style-id:electionLink" href="http://www.cnn.com/election/house">House</a> | <a class="style-id:electionLink" href="http://www.cnn.com/election/governor">Governer</a> | <a class="style-id:electionLink" href="http://www.cnn.com/election/ballot-measures">Ballot&nbsp;Measures</a> | <a class="style-id:electionLink" href="http://www.cnn.com/election/results/exit-polls">Exit&nbsp;Polls</a> | <a class="style-id:electionLink" href="http://www.cnn.com/election/results/states">States</a></p>',
            gnsElectiomImgEnv: (nconf.get('GNS_ELECTION_IMG_ENV')) ? nconf.get('GNS_ELECTION_IMG_ENV') : nconf.get('ENVIRONMENT'),
            gnsEnableInlineGalleries: (nconf.get('GNS_ENABLE_INLINE_GALLERIES')) ? nconf.get('GNS_ENABLE_INLINE_GALLERIES') : false,
            gnsMonitoringTest: (nconf.get('GNS_MONITORING_TEST')) ? nconf.get('GNS_MONITORING_TEST') : false,
            gnsGenericThumbImage: (nconf.get('GNS_GENERIC_THUMB_IMAGE')) ? nconf.get('GNS_GENERIC_THUMB_IMAGE') : 'http://www.cnn.com/partners/google/gns/default-exlarge-169.png',
            hypatia: {
                timeout: (process.env.HYPATIA_TIMEOUT) ? parseInt(process.env.HYPATIA_TIMEOUT) : 1000 * 5,
                host: (process.env.HYPATIA_HOST) ? process.env.HYPATIA_HOST :  'https://hypatia.api.cnn.com/'
            },
            aws: {
                accessKeyId: nconf.get('ACCESS_KEY_ID'),
                secretAccessKey: nconf.get('SECRET_ACCESS_KEY'),
                region: 'us-east-1',
                bucket: 'registry.api.cnn.io'
            },
            logConfig: (typeof process.env.CUSTOMER === 'undefined') ? null : {logzio: {tag: `cnn-google-newsstand-${nconf.get('ENVIRONMENT').toLowerCase()}`}},
            sections: {
                business: nconf.get('BUSINESS_SECTIONS') ? nconf.get('BUSINESS_SECTIONS') : '\/business|success|india|media|perspectives|investing|cars|economy|energy|homes\/'
            }
        },
        prod: {
            cloudamqpConnectionString: `amqp://${nconf.get('CLOUDAMQP_AUTH')}@red-rhino.rmq.cloudamqp.com/cnn-towncrier`,
            exchangeName: 'cnn-town-crier-prod',
            queueNameArticles: 'cnn-google-newsstand-articles-prod',
            queueNameVideos: 'cnn-google-newsstand-videos-prod',
            adbpTrackingURL: 'https://smetrics.cnn.com/b/ss/cnn-adbp-offsite-domestic'
        }
    };


// load the correct config based on environment
switch (nconf.get('ENVIRONMENT').toLowerCase()) {
    case 'prod':
        nconf.defaults(config.prod);
        break;


    default:
        nconf.defaults(config.default);
}



// Load overrides that don't override anything, they fill in the blanks
nconf.overrides(config.default);



module.exports = nconf;


server.js
/*
 * Copyright 2016 Turner Broadcasting System, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

const hapi = require('cnn-hapi'),
    config = require('./config.js'),
    pkg = require('./package.json'),
    healthCheck = require('./tasks/healthcheck.js');

global.gnsHealthStatus = {
    status: 201,
    testMode: config.get('gnsMonitoringTest'),
    sectionFeeds: {
        business: {
            status: 201,
            generateFeed: {status: 'Pending'}
        },
        us: {
            status: 201,
            generateFeed: {status: 'Pending'}
        },
        world: {
            status: 201,
            generateFeed: {status: 'Pending'}
        },
        politics: {
            status: 201,
            generateFeed: {status: 'Pending'}
        },
        entertainment: {
            status: 201,
            generateFeed: {status: 'Pending'}
        },
        tech: {
            status: 201,
            generateFeed: {status: 'Pending'}
        },
        latest: {
            status: 201,
            generateFeed: {status: 'Pending'}
        },
        // TODO: Remove once business feed is confirmed in production
        money: {
            status: 201,
            generateFeed: {status: 'Pending'}
        },
        health: {
            status: 201,
            generateFeed: {status: 'Pending'}
        },
        opinions: {
            status: 201,
            generateFeed: {status: 'Pending'}
        },
        videos: {
            status: 201,
            generateFeed: {status: 'Pending'}
        },
        elections: {
            status: 201,
            generateFeed: {status: 'Pending'}
        },
        style: {
            status: 201,
            generateFeed: {status: 'Pending'}
        },
        travel: {
            status: 201,
            generateFeed: {status: 'Pending'}
        },
        winterOlympics2018: {
            status: 201,
            generateFeed: {status: 'Pending'}
        }
    }
};

let server = module.exports = hapi({
    directory: __dirname,
    port: process.env.PORT,
    withSwagger: true,
    name: 'CNN Google Newsstand',
    description: 'Google Newsstand feed generator',
    version: pkg.version
});

process.on('unhandledRejection', function (error, promise) {
    console.error(`Possible unhandled rejection at: Promise ${JSON.stringify(promise)} reason: ${error.stack}`);
});

server.route({
    method: 'GET',
    path: '/',
    handler: function (request, reply) {
        reply.redirect('/documentation');
    }
});

server.route({
    method: 'GET',
    path: '/_healthcheck',
    handler: function healthcheckHandler(request, reply) {
        let checks = healthCheck.getStatus();

        checks.version = pkg.version;
        reply(checks);
    },
    config: {
        description: 'Healthcheck',
        notes: 'Health of the app',
        tags: ['api', 'healthcheck']
    }
});

server.start(function () {
    console.log(`Server running at ${JSON.stringify(server.info.uri)}`);
    console.log('Configuration:');
    console.log(`    ENVIRONMENT: ${config.get('ENVIRONMENT')}`);
    console.log(`    PORT: ${config.get('PORT')}`);
    console.log(`    GNS_TASK_INTERVAL_MS: ${config.get('GNS_TASK_INTERVAL_MS')}`);

    require('./tasks/google-newsstand-articles.js');
    require('./tasks/google-newsstand-videos.js');
});


tasks/generate-static-feed.js
/*
 * Copyright 2016 Turner Broadcasting System, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

 /*
  * This is a debugging task to post specific articles to a specific endpoint.
  *
  * $ ENVIRONMENT=local PORT=5000 CLOUDAMQP_AUTH=foo node tasks/generate-static-feed.js
  *
  * You can also manually capture the feed and push it with:
  *
  * $ for dc in cop 56m; do curl -H "Content-Type: application/rss+xml" --data @test.ig.test/test9.xml lsd-prod-pub-${dc}.turner.com/cnn/content/google-newsstand/test6.xml; done
  */

'use strict';

const  request = require('request'),
    AWS = require('aws-sdk'),
    _ = require('underscore'),
    maxKeys = 1000,
    FeedGenerator = require('../lib/feed-generator.js'),
    debugLog = require('debug')('cnn-google-newsstand:Task:generate-static-feed'),
    config = require('../config.js'),
    fg = new FeedGenerator(),
    enableElectionStory = config.get('gnsTurnOnElectionModule'),
    logConfig = config.get('logConfig'),
    log = require('cnn-logger')(logConfig),
    POST_TO_LSD = false; // <---- TODO - SET THIS TO THE PROPER VALUE BASED ON WHAT YOU ARE WANTING TO DO

let s3Images = undefined;

function postToLSD(data) {

    let suffix = (config.get('ENVIRONMENT') === 'prod') ? '' : `-${config.get('ENVIRONMENT')}`,
        endpoint = `/cnn/content/google-newsstand/test-not-public-inline-gallery${suffix}.xml`,  // <---- TODO - SET THIS TO THE CORRECT ENDPOINT BEFORE RUNNING
        hosts = config.get('lsdHosts');

    debugLog('postToLSD() called');
    log.debug('postToLSD() called');

    // debugLog(data);

    hosts.split(',').forEach((host) => {
        request.post({
            url: `http://${host}${endpoint}`,
            body: data,
            headers: {'Content-Type': 'application/rss+xml'}
        },
        (error/* , response, body*/) => {
            if (error) {
                debugLog(error.stack);
                log.error(error.stack);
            } else {
                debugLog(`Successfully uploaded data to ${hosts} at ${endpoint}`);
                log.debug(`Successfully uploaded data to ${hosts} at ${endpoint}`);
                // debugLog(body);
            }
        });
    });
}

function filterImages(data) {
    let filteredImages = [],
        contents = data,
        image,
        i;
    for (i in contents) {
        image = contents[i];
        if (image.Key.includes('.png')) {
            filteredImages.push({
                time: image.LastModified,
                url: `http://registry.api.cnn.io/${image.Key}`
            });
        }
    }

    filteredImages = _.sortBy(filteredImages, function (o) {
        return new Date(o.time).getTime();
    });

    return filteredImages;
}

/**
 * List keys from the specified bucket.
 *
 * If providing a prefix, only keys matching the prefix will be returned.
 *
 * If providing a delimiter, then a set of distinct path segments will be
 * returned from the keys to be listed. This is a way of listing "folders"
 * present given the keys that are there.
 *
 * @param {Object} options
 * @param {String} options.bucket - The bucket name.
 * @param {String} [options.prefix] - If set only return keys beginning with
 *   the prefix value.
 * @param {String} [options.delimiter] - If set return a list of distinct
 *   folders based on splitting keys by the delimiter.
 * @param {Function} callback - Callback of the form function (error, string[]).
 */
function listKeys(options, callback) {
    var keys = [];

    /**
    * Recursively list keys.
    *
    * @param {String|undefined} marker - A value provided by the S3 API
    *   to enable paging of large lists of keys. The result set requested
    *   starts from the marker. If not provided, then the list starts
    *   from the first key.
    */
    function listKeysRecusively(marker) {

        options.marker = marker;

        listKeyPage(options, function (error, nextMarker, keyset) {
            if (error) {
                return callback(error, keys);
            }

            keys = keys.concat(keyset);

            if (nextMarker) {
                listKeysRecusively(nextMarker);
            } else {
                callback(null, keys);
            }
        });
    }

    // Start the recursive listing at the beginning, with no marker.
    listKeysRecusively();
}

/**
 * List one page of a set of keys from the specified bucket.
 *
 * If providing a prefix, only keys matching the prefix will be returned.
 *
 * If providing a delimiter, then a set of distinct path segments will be
 * returned from the keys to be listed. This is a way of listing "folders"
 * present given the keys that are there.
 *
 * If providing a marker, list a page of keys starting from the marker
 * position. Otherwise return the first page of keys.
 *
 * @param {Object} options
 * @param {String} options.bucket - The bucket name.
 * @param {String} [options.prefix] - If set only return keys beginning with
 *   the prefix value.
 * @param {String} [options.delimiter] - If set return a list of distinct
 *   folders based on splitting keys by the delimiter.
 * @param {String} [options.marker] - If set the list only a paged set of keys
 *   starting from the marker.
 * @param {Function} callback - Callback of the form
    function (error, nextMarker, keys).
 */
function listKeyPage(options, callback) {
    let params = {
            Bucket: options.bucket,
            Delimiter: options.delimiter,
            Marker: options.marker,
            MaxKeys: maxKeys,
            Prefix: options.prefix
        },
        s3Client,
        awsConfig = {accessKeyId: config.get('aws').accessKeyId, secretAccessKey: config.get('aws').secretAccessKey, region: 'us-east-1'};


    AWS.config.update(awsConfig);
    s3Client = new AWS.S3();

    s3Client.listObjects(params, function (error, response) {
        if (error) {
            return callback(error);
        } else if (response.err) {
            return callback(new Error(response.err));
        }

        // Convert the results into an array of key strings, or
        // common prefixes if we're using a delimiter.
        var keys, nextMarker, lastKey;
        if (options.delimiter) {
          // Note that if you set MaxKeys to 1 you can see some interesting
          // behavior in which the first response has no response.CommonPrefix
          // values, and so we have to skip over that and move on to the
          // next page.
            keys = _.map(response.CommonPrefixes, function (item) {
                return item.Prefix;
            });
        } else {
            keys = _.map(response.Contents, function (item) {
                return item;
            });
        }

        // Check to see if there are yet more keys to be obtained, and if so
        // return the marker for use in the next request.
        if (response.IsTruncated) {
            if (options.delimiter) {
                // If specifying a delimiter, the response.NextMarker field exists.
                nextMarker = response.NextMarker;
            } else {
                // For normal listing, there is no response.NextMarker
                // and we must use the last key instead.
                lastKey = keys[keys.length - 1];


                nextMarker = lastKey.Key;
            }
        }

        callback(null, nextMarker, keys);
    });
}


function getImagesFromAWS() {
    return new Promise(function (fulfill) {

        listKeys({
            bucket: config.get('aws').bucket,
            prefix: `assets/img/opp/ksa/${config.get('gnsElectiomImgEnv')}/gns/`
        }, function (error, keys) {
            if (error) {
                console.log('Error retrieving images from s3', error);
                log.error(`Error retrieving images from s3': ${error}`);
                fulfill({error: 'Error retrieving images from s3'});
            }

            console.log('Successfully retrieved images from s3, about to fulfill.. keys.length: ', keys.length);
            log.debug(`Successfully retrieved images from s3, about to fulfill.. keys.length: ' ${keys.length}`);
            fulfill(filterImages(keys));
        });
    });
}

function isConstantPublishedAlreadyThere(urls, electionStoryUrl) {
    let isStoryThere = false;

    urls.some((url) => {
        if (electionStoryUrl === url) {
            isStoryThere = true;
        }
    });

    return isStoryThere;
}

if (enableElectionStory === true || enableElectionStory === 'true') {
    s3Images = getImagesFromAWS();
}

// put URLs you want to test against here
fg.urls = [
    // 'http://www.cnn.com/2017/06/09/design/gallery/thom-browne/index.html'
    // 'http://www.cnn.com/travel/article/frank-lloyd-wright-turns-150-architecture-tour-us-midwest/index.html'//  page top gallery
    // 'http://www.cnn.com/2017/06/16/entertainment/tupac-all-eyez-on-me/index.html' // page top gallery
    // 'http://www.cnn.com/2014/01/25/showbiz/gallery/bachelor-where-are-they-now/index.html' // stand alone gallery
    // 'http://www.cnn.com/2016/08/08/opinions/mcmullin-mormon-hope-for-conservatives-stanley/index.html'
    // 'http://www.cnn.com/2016/08/08/sport/aly-raisman-parents-olympics-trnd/index.html' // page top image
    // 'http://www.cnn.com/2016/08/08/sport/office-olympics-for-the-rest-of-us-trnd/index.html' // page top video
    // 'http://www.cnn.com/2016/08/08/sport/aly-raisman-parents-olympics-trnd/index.html' // page top image
    // 'http://www.cnn.com/2016/08/08/sport/office-olympics-for-the-rest-of-us-trnd/index.html', // page top video
    // 'http://www.cnn.com/2016/08/09/health/parent-acts-teaching-kids-empathy/index.html' // page top video collection
    // 'http://www.cnn.com/2016/07/27/us/freddie-gray-verdict-baltimore-officers/index.html' // page top video collection / inline video / inline images
    // 'http://www.cnn.com/2016/08/15/opinions/is-trump-getting-ready-to-lose-stanley/index.html'
    // 'http://www.cnn.com/2016/08/15/health/parents-life-expectancy-heart-health/index.html' //
    // 'http://www.cnn.com/2016/08/10/politics/trump-second-amendment/index.html'//
    // 'http://www.cnn.com/2016/07/22/architecture/leaning-house-jakarta-architecture/index.html',
    // 'http://www.cnn.com/2016/07/19/sport/jockey-horse-diets/index.html',
    // 'http://www.cnn.com/2016/07/18/travel/national-seashore-lakeshore-towns-nps100/index.html'
    // 'http://www.cnn.com/2016/08/15/us/gabby-douglas-natalie-hawkins-new-day/index.html', // twitter embeds
    // 'http://www.cnn.com/2015/10/13/politics/democratic-debate-2016-instagram/index.html' // ig embeds
    // 'http://www.cnn.com/2016/09/05/hotels/presidential-hotel-suites/index.html' // page top gallery
    // 'http://www.cnn.com/2016/08/16/opinions/larry-wilmore-cancellation-obeidallah/index.html' // editors note  / image / video
    // 'http://www.cnn.com/2016/08/19/entertainment/amy-schumer-charlie-rose/index.html', // youtube
    // 'http://www.cnn.com/2015/10/14/politics/democratic-debate-in-gifs-vines-clinton-sanders-reaction/index.html' // vine
    // 'http://www.cnn.com/2016/01/05/politics/primary-conflicts-2016/index.html' // vimeo
    // videos
    // 'http://www.cnn.com/videos/politics/2016/09/20/who-stole-naked-trump-moos-pkg-erin.cnn'
    // 'http://www.cnn.com/videos/us/2016/09/01/midway-atoll-plastic-island-obama-visits-wildlife-refuge-marine-reserve-orig.cnn'
    // 'http://www.cnn.com/videos/entertainment/2016/08/30/chance-the-rapper-beyonce-mtv-vma-orig-vstan.cnn',
    // 'http://www.cnn.com/videos/us/2016/09/01/houston-crosswalks-vandalized-pkg.ktrk'
    // 'http://www.cnn.com/2016/09/20/politics/gun-control-law-court/index.html'
    // 'http://www.cnn.com/2016/09/26/opinions/clinton-needs-obama-bernstein-opinion/index.html'
    // 'http://www.cnn.com/2014/06/09/world/boko-haram-fast-facts/index.html'
    // 'http://www.cnn.com/videos/us/2016/09/01/houston-crosswalks-vandalized-pkg.ktrk'
    // 'http://www.cnn.com/2016/09/20/politics/gun-control-law-court/index.html',
    // 'http://ref.next.cnn.com/2016/10/12/politics/test-election-day-2016-highlights/index.html'
    // 'http://www.cnn.com/2016/09/26/opinions/clinton-needs-obama-bernstein-opinion/index.html',
    // 'http://www.cnn.com/2016/10/28/politics/donald-trump-election-2016/index.html'
];

if (fg.urls && fg.urls.length > 0) {


    if (s3Images) {

        console.log('s3images = true');
        log.debug('s3images = true');

        s3Images.then(function (data) {

            let constantElectionStoryUpdate = config.get('gnsElectionStoryConstantUpdate'),
                constantElectionStoryUpdateURL = config.get('gnsElectionStoryConstantUpdateURL');

            if ((constantElectionStoryUpdate === 'true' || constantElectionStoryUpdate === true) && constantElectionStoryUpdateURL) {
                if (!isConstantPublishedAlreadyThere(fg.urls, constantElectionStoryUpdateURL)) {
                    fg.urls.unshift(constantElectionStoryUpdateURL);
                }
            }

            fg.processContent({s3Data: data}).then(
                // success
                (rssFeed) => {
                    console.log(rssFeed);

                    if (POST_TO_LSD) {
                        postToLSD(rssFeed);
                    }

                    // post to LSD endpoint
                    fg.urls = 'clear';
                },

                // failure
                (error) => {
                    console.log(`Error: ${error}`);
                    log.error(error);
                }
            );
        });
    } else {

        console.log('s3images = false');
        log.debug('s3images = false');

        fg.processContent().then(
            // success
            (rssFeed) => {
                console.log(rssFeed);

                if (POST_TO_LSD) {
                    postToLSD(rssFeed);
                }

                // post to LSD endpoint
                fg.urls = 'clear';
            },

            // failure
            (error) => {
                console.log(`Error: ${error}`);
                log.error(error);
            }
        );
    }
} else {
    debugLog('no updates');
    log.debug('no updates');
}
